refactor(vitest): reimplement hoistMocks to fix incorrect dynamic imports

Dunqing opened the pull request:
### Description

fix: #4619 
fix: #4604
fix: #3300

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

<!-- You can also add additional context here -->

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [x] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

sheremet-va commented on packages/browser/rollup.config.js:
```diff
@@ -14,6 +14,7 @@ const external = [
   /^@?vitest(\/|$)/,
   'worker_threads',
   'node:worker_threads',
+  '@vitest/utils/ast',
```
It should already be covered by a regexp on L14

sheremet-va commented on packages/vitest/src/node/hoistMocks.ts:
```diff
@@ -71,58 +72,114 @@ export function hoistMocks(code: string, id: string, parse: PluginContext['parse
+      if (!binding)
+        return
+
+      if (isStaticProperty(parent) && parent.shorthand) {
```
I feel like these if statements are duplicated here and in browser. Maybe insteadof `parentStack` we can pass down some kind of options object? `hasBindingShortcut: true, classDeclaration: false, classExpression: false` (don't forget to add comments with example why we are doing g the transform like in property shorthand)

Dunqing replied:
Sounds good

sheremet-va commented on packages/vitest/src/node/hoistMocks.ts:
```diff
@@ -71,58 +72,114 @@ export function hoistMocks(code: string, id: string, parse: PluginContext['parse
+    onImportMeta() {},
+  })
+
   simpleWalk(ast, {
```
Can we include that in a previous walk for performance reasons?

Dunqing replied:
Yes, Let me make some changes in `esmWalker`

sheremet-va commented on test/core/test/spy.test.ts:
```diff
@@ -1,5 +1,7 @@
 import { describe, expect, test, vi } from 'vitest'
 import * as mock from './fixtures/hello-mock'
+import * as squaredModule from './fixtures/squared'
```
We need more tests for this behaviour. I also added two more issues in your description - you can look it up there. The tests should be in examples/mocks

sheremet-va commented on test/core/test/injector-mock.test.ts:
```diff
@@ -67,7 +67,6 @@ test('correctly mocks namespaced', () => {
     "const { vi } = await import('vitest')
     vi.mock('../src/add', () => {})
-    const AddModule = await import('../src/add')
-    const { default: add } = AddModule"
```
There are no tests that check if properties are accessed correctly. Maybe we can take them from Vite? This function originated there

Dunqing replied:
It doesn't look like we need to do this. The current behavior also works fine

sheremet-va replied:
No, we _need_ to have tests.

Dunqing replied:
I checked in vite. Looks like we don't need this line `const { default: add } = AddModule"
`
```ts
  expect(await ssrTransformSimpleCode(`
    import { vi } from 'vitest'
    import add, * as AddModule from '../src/add'
    add();
    vi.mock('../src/add', () => {})`))
.toMatchInlineSnapshot(`
    "const __vite_ssr_import_0__ = await __vite_ssr_import__("vitest", {"importedNames":["vi"]});
    const __vite_ssr_import_1__ = await __vite_ssr_import__("../src/add", {"importedNames":["default"]});

      
      
      __vite_ssr_import_1__.default();
      __vite_ssr_import_0__.vi.mock('../src/add', () => {})"
  `)
```

sheremet-va replied:
Then add these tests in this pr

sheremet-va commented:
I cannot accept this PR without tests. Vitest is a testing framework and the lack of tests for our own features is not a good sign. Initially, this was an experimental feature to test how it works, but now it does more than before. Since it doesn't use Vite's SSR transformer directly, we need to have tests in our repo. You can just copy SSR transformer tests from Vite.

Dunqing commented:
> I cannot accept this PR without tests. Vitest is a testing framework and the lack of tests for our own features is not a good sign. Initially, this was an experimental feature to test how it works, but now it does more than before. Since it doesn't use Vite's SSR transformer directly, we need to have tests in our repo. You can just copy SSR transformer tests from Vite.

Sorry! I finally got what you meant. I will add these tests later

Dunqing commented:
I added tests and made some changes. Since the `hoistMocks` must the code has mocks code. so I added some mock code combined with the original code. Please feel free to point out if it's not right!

sheremet-va commented on test/core/test/injector-mock.test.ts:
```diff
@@ -67,7 +72,1130 @@ test('correctly mocks namespaced', () => {
+
+describe('transform', () => {
+  const hoistSimpleCodeWithoutMocks = (code: string) => {
+    return hoistMocks(`import {vi} from "vitest";\nvi.mock(faker);\n${code}`, '/test.js', parse)?.code.trim()
```
using `faker` without any reference to it feels bad. Maybe you can just use a string identifier?

Dunqing replied:
Sorry. I don't understand the second one. Do you mean I should put `vi.mock` after the original code?

Dunqing replied:
I seem to understand what you mean, I've changed to string literal. this test code makes more sense

sheremet-va commented on test/core/test/injector-mock.test.ts:
```diff
@@ -67,7 +72,1130 @@ test('correctly mocks namespaced', () => {
+      vi.mock(faker)
+      const __vi_import_0__ = await import('vue')
+
+      ;
```
I know that it was before your refactor, but this hanging `;` feels lonely, maybe we can remove it somehow?

sheremet-va replied:
Ok

Dunqing replied:
This is an easy fix, I've already committed the code 😅

sheremet-va commented on test/core/test/injector-mock.test.ts:
```diff
@@ -67,7 +72,1149 @@ test('correctly mocks namespaced', () => {
+      vi.mock('../src/add', () => {})`),
+    ).toMatchInlineSnapshot(`
+      "const { vi } = await import('vitest')
+      const { vi } = await import('vitest')
```
This test is a bit weird. It was here before the refactor and it tested that the default export was correctly extracted

Right now, since `add` and `AddModule` are not used anywhere it looks redundant. Also the name "test" for a test is not telling much

Dunqing replied:
I'm sorry, I added this test case to test certain scenarios. I forgot to remove it.

sheremet-va commented:
@Dunqing this broke tests using imported variables inside the factory: #4872
PR with a failing test at #4873
